crates/hir-ty/src/tests/traits.rs

```diff
@@ -3963,3 +3963,160 @@ fn g(t: &(dyn T + Send)) {
         "#,
     );
 }
+
+#[test]
+fn gats_in_path() {
+    check_infer_with_mismatches(
+        r#"
+//- minicore: deref
+use core::ops::Deref;
+trait PointerFamily {
+    type Pointer<T>: Deref<Target = T>;
+}
+
+fn f<P: PointerFamily>(p: P::Pointer<i32>) {
+    let a = *p;
+}
+fn g<P: PointerFamily>(p: <P as PointerFamily>::Pointer<i32>) {
+    let a = *p;
+}
+        "#,
+        expect![[r#"
```

Comment:
in general, I want to get rid of the expect tests and move to `check` etc., because they're more flexible and it's a bit clearer what we're actually testing (vs. what's just incidental code)